### PR TITLE
Adding more functions of adb and refactoring tests

### PR DESCRIPTION
**Pull request by moizjv**
- Refactored tests added functional and unit tests
- Added more methods and tests to ADB. 
- Unit tests are still work in progress, aiming to achieve more coverage.

### CODE REVIEW

File: lib/tools/adb-commands.js [hunk 1]
```diff
@@ -14,4 +16,97 @@ methods.initZipAlign = async function() {
...
+    let engines = [];
+    for (let line of stdout.split('\n')) {
+      if (line.length > 0 && line[0] !== ' ') {
+        // remove newline and trailing colon, and add to the list
+        engines.push(line.trim().replace(/:$/, ''));
+      }
+    }
+    return engines;
+  } catch (e) {
+    throw new Error(`Error getting available IME's. Original error: ${e.message}`);
```
Comment (jlipps) [inline lib/tools/adb-commands.js]: in all of these places where we're throwing errors: in the current code, are we also logging the errors? or no

File: lib/tools/adb-commands.js [hunk 2]
```diff
@@ -14,4 +16,97 @@ methods.initZipAlign = async function() {
...
+  }
+};
+
+methods.enabledIMEs = async function () {
```
Comment (jlipps) [inline lib/tools/adb-commands.js]: lot of duplication between this and the previous function, suggests a factoring out of something

File: lib/tools/adb-commands.js [hunk 3]
```diff
@@ -14,4 +16,97 @@ methods.initZipAlign = async function() {
...
+      }
+    }
+    return engines;
+  } catch (e) {
+    throw new Error(`Error getting available IME's. Original error: ${e.message}`);
+  }
+};
+
+methods.enableIME = async function (imeId) {
+  return this.shell(['ime', 'enable', imeId]);
```
Comment (jlipps) [inline lib/tools/adb-commands.js]: are there actual return values of these ime methods? if not, I would probably write it as this instead:

```js
methods.enableIME = async function (imeId) {
  await this.shell(['ime', 'enable', imeId]);
};
```

since it makes it more semantically transparent that what we're doing is simply waiting for the action to complete

File: lib/tools/adb-commands.js (hunk 2 above)
Comment (moizjv) [inline lib/tools/adb-commands.js]: Yes this was actually a after thought, this function was not present and I added it, becuase i see a value in it. Yes will refactor :+1:

File: lib/tools/system-calls.js [hunk 4]
```diff
@@ -75,4 +76,112 @@ systemCallMethods.getBinaryFromPath = async function(binaryName) {
...
+    try {
+      let devices = await this.getConnectedDevices();
+      if (devices.length < 1) {
+        log.debug("Could not find devices, restarting adb server...");
+        await this.restartAdb();
+        // cool down
+        await sleep(200);
+        await getDevices();
+      }
+      return devices;
```
Comment (jlipps) [inline lib/tools/system-calls.js]: it looks like we return this `devices` value regardless if whether we do the restart in the `if` block above. if that's the case, then why is the `if` block even there? seems like we should be reassigning `devices` inside that block somewhere...

File: lib/tools/system-calls.js [hunk 5]
```diff
@@ -75,4 +76,112 @@ systemCallMethods.getBinaryFromPath = async function(binaryName) {
...
+      await exec(this.adb.path, ['kill-server']);
+    } catch (e) {
+      log.error("Error killing ADB server, going to see if it's online anyway");
+    }
+  }
+};
+
+systemCallMethods.adbExec = async function (cmd, opts = {}) {
+  if (!cmd) {
+    throw new Error("You need to pass in a command to exec()");
```
Comment (jlipps) [inline lib/tools/system-calls.js]: maybe change this to "adbExec"?

File: lib/tools/adb-commands.js (hunk 3 above)
Comment (moizjv) [inline lib/tools/adb-commands.js]:
```
ADB.prototype.enableIME = function (imeId, cb) {
  var cmd = 'ime enable ' + imeId;
  this.shell(cmd, cb);
};
```

This was the earlier implementation. And output looks like `Input method com.example.android.softkeyboard/.SoftKeyboard: already enabled`

File: lib/tools/system-calls.js (hunk 5 above)
Comment (moizjv) [inline lib/tools/system-calls.js]: copy paste curse will change it

File: test/functional/syscalls-e2e-specs.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    devices.should.have.length.above(0);
+  });
+  it('adbExec should get devices when with devices', async () => {
+    (await adb.adbExec("devices")).should.contain("List of devices attached");
+  });
+  it('isDeviceConnected should be true', async () => {
+    (await adb.isDeviceConnected()).should.be.true;
+  });
+  it('shell should execute command in adb shell ', async () => {
+    (await adb.shell(['getprop', 'ro.build.version.sdk'])).should.be.equal(apiLevel);
```
Comment (jlipps) [inline test/functional/syscalls-e2e-specs.js]: instead of `should.be.equal(` let's do `should.equal(` in any place that has it. it's more grammatical :-)

File: test/unit/adb-commands-specs.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    afterEach(async () => {
+      adb.shell.restore();
+    });
+    let verifyShellArguments = function (cmd) {
+      (shell.args[0][0]).should.be.deep.equal(cmd);
+    };
+    it('getApiLevel should call shell with correct args', async () => {
+      createStub('21');
+      (await adb.getApiLevel()).should.be.equal(apiLevel);
+      verifyShellArguments([ 'getprop', 'ro.build.version.sdk' ]);
```
Comment (jlipps) [inline test/unit/adb-commands-specs.js]: no need for spaces here

**Comment by jlipps** [comment]
ok, finished review and made comments. overall looks really great! i do like the unit tests and think we should keep going in this direction.

File: lib/tools/system-calls.js (hunk 4 above)
Comment (moizjv) [inline lib/tools/system-calls.js]: this is little tricky i had an `else { return devices }`. But actually it will never return as we recursively call `await getDevices()`

File: lib/tools/adb-commands.js (hunk 3 above)
Comment (jlipps) [inline lib/tools/adb-commands.js]: I mean, does it make sense to be returning a random stdout string from this Node-js based API? Thinking about what kind of API we're creating a few things make sense to me:

1. returning nothing (it's just an action after all)
2. returning true or false based on whether the action actually took place
3. throwing an error if the action didn't complete successfully

but returning a stdout string seems a little wonky to me.

File: lib/tools/system-calls.js (hunk 4 above)
Comment (jlipps) [inline lib/tools/system-calls.js]: but just calling `await getDevices()` recursively doesn't get that value back to the original calling function ever, right? this probably needs some heavy unit tests since there's so much actual logic. on first pass I think that would show up a bug. I think this block should probably look like:

```js
      let devices = await this.getConnectedDevices();
      if (devices.length < 1) {
        log.debug("Could not find devices, restarting adb server...");
        await this.restartAdb();
        // cool down
        await sleep(200);
        devices = await getDevices();
      }
      return devices;
```

but i could be wrong

File: test/unit/adb-commands-specs.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    });
+  };
+  beforeEach(async () => {
+    await adb.createADB();
+  });
+  describe('shell', () => {
+    afterEach(async () => {
+      adb.shell.restore();
+    });
+    let verifyShellArguments = function (cmd) {
```
Comment (sebv) [inline test/unit/adb-commands-specs.js]: You probably want to use a  a mock for this kind of stuff.

File: test/unit/adb-commands-specs.js [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+      (shell.args[0][0]).should.be.deep.equal(cmd);
```
Comment (sebv) [inline test/unit/adb-commands-specs.js]: I think you can remove `be`

File: test/unit/adb-commands-specs.js [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+      (await adb.getApiLevel()).should.be.equal(apiLevel);
+      verifyShellArguments([ 'getprop', 'ro.build.version.sdk' ]);
+    });
+    it('availableIMEs should call shell with correct args', async () => {
+      createStub(imeList);
+      (await adb.availableIMEs()).should.have.length.above(0);
+      verifyShellArguments(['ime', 'list', '-a']);
+    });
+    it('enabledIMEs should call shell with correct args', async () => {
+      createStub(imeList);
```
Comment (sebv) [inline test/unit/adb-commands-specs.js]: I would try to avoid setting up stub/mock inside `it` block.

File: lib/tools/system-calls.js [hunk 11]
```diff
@@ -75,4 +76,112 @@ systemCallMethods.getBinaryFromPath = async function(binaryName) {
   }
 };
 
+systemCallMethods.getConnectedDevices = async function () {
+  log.debug("Getting connected devices...");
+  try {
+    let {stdout} = await exec(this.adb.path, ['devices']);
```
Comment (imurchie) [inline lib/tools/system-calls.js]: This is just out of curiosity, but what is up with `let {...} = ...`?

Comment (jlipps) [inline lib/tools/system-calls.js]: @imurchie this is for convenience. `teen_process.exec()` returns an object that looks like:

```js
{
  stdout: 'some string',
  stderr: 'some string',
  code: 0
}
```

Instead of returning an object, the other option was to return an array. Then we could have retrieved the values like:

```js
let [stdout, stderr, code] = await exec(...);
```

But in many cases we want only one type of thing from `exec`. Let's say we only wanted `stderr`, then it would look like:

```js
let [, stderr] = await exec(...);
```

Which is slightly more mysterious than what we can do now (because of object destructuring):

```js
let {stderr} = await exec(...);
```

In other words, returning an object plus ES6 object destructuring gives us the ability to target simply those return values we're interested in.

Comment (imurchie) [inline lib/tools/system-calls.js]: Ah! Thanks. I knew about object destructuring, but didn't clue in to the pattern here.

File: lib/tools/adb-commands.js (hunk 1 above)
Comment (moizjv) [inline lib/tools/adb-commands.js]: We are currently now logging errors, I am not sure if want to log error messages?

File: test/unit/adb-commands-specs.js (hunk 10 above)
Comment (moizjv) [inline test/unit/adb-commands-specs.js]: Each test case requires different output. I am not sure how I can avoid creating a new stub every time to have a different output?

File: test/unit/adb-commands-specs.js (hunk 8 above)
Comment (moizjv) [inline test/unit/adb-commands-specs.js]: sinon does not allow me to mock already stubbed function 

`TypeError: Attempted to wrap shell which is already wrapped`

File: lib/tools/adb-commands.js (hunk 1 above)
Comment (jlipps) [inline lib/tools/adb-commands.js]: the nice thing about logging them where we throw them is we get to see what happened just before. not just the overall failure without an idea of where in the log flow it actually took place

**Comment by moizjv** [comment]
@jlipps @sebv made more changes, going to add logging in `catch (e)` cases

File: lib/tools/adb-commands.js [hunk 12]
```diff
@@ -14,4 +17,86 @@ methods.initZipAlign = async function() {
...
+methods.clear = function (pkg) {
+  return this.shell(['pm', 'clear', pkg]);
+};
+
+methods.stopAndClear = async function (pkg) {
+  try {
+    await this.forceStop(pkg);
+    await this.clear(pkg);
+  } catch (e) {
+    log.error(`Cannot stop and clear ${pkg}. Original error: ${e.message}`);
```
Comment (jlipps) [inline lib/tools/adb-commands.js]: let's not duplicate the error message. Instead, we should do:

```js
let msg = `Cannot stop and clear ${pkg}. Original error: ${e.message}`;
log.error(msg);
throw new Error(msg);
```

Comment (jlipps) [inline lib/tools/adb-commands.js]: we might want to add a function `logAndThrow(msg)` to appium-support since this is a common pattern. then you could just do:

```js
logAndThrow(`Cannot stop and clear ${pkg}. Original error: ${e.message}`);
```

Comment (jlipps) [inline lib/tools/adb-commands.js]: or we could add it as a special method on our appium-logger object, so you could do:

```js
log.throw(`Cannot stop and clear ${pkg}. Original error: ${e.message}`);
```

**Comment by moizjv** [comment]
Also added unit test cases for recursion `getDevicesWithRetry` :)

File: test/unit/adb-commands-specs.js (hunk 8 above)
Comment (sebv) [inline test/unit/adb-commands-specs.js]: Yeah it's either stub or mock, a mock is a stub + expectations.

Comment (moizjv) [inline test/unit/adb-commands-specs.js]: then i will have to stick to stub as i am mocking shell return value

File: test/unit/adb-commands-specs.js (hunk 10 above)
Comment (sebv) [inline test/unit/adb-commands-specs.js]: For instance:
```js
describe(''enabledIMEs', () => {
    before(() = {  createStub(imeList);  });
     it('should call shell with correct args', async', () => {
         ...
     });
     after(() => { clearStub() })
});
```

Comment (moizjv) [inline test/unit/adb-commands-specs.js]: Ohh I see what you are saying.

File: test/unit/adb-commands-specs.js (hunk 8 above)
Comment (sebv) [inline test/unit/adb-commands-specs.js]: btw there are good examples here http://dailyjs.com/2013/02/14/backbone-tutorial-12/
https://www.dropbox.com/s/xft7u9ub4rpy93a/Screenshot%202015-06-03%2009.00.13.png?dl=0

File: lib/tools/system-calls.js [hunk 13]
```diff
@@ -70,9 +71,127 @@ systemCallMethods.getBinaryFromPath = async function(binaryName) {
...
+      log.debug(`${devices.length} device(s) connected`);
+      return devices;
+    }
+  } catch (e) {
+    log.error(`Error while getting connected devices. Original error: ${e.message}`);
+    throw new Error(`Error while getting connected devices. Original error: ${e.message}`);
+  }
+};
+
+systemCallMethods.getDevicesWithRetry = async function (timeoutMs = 20000) {
```
Comment (sebv) [inline lib/tools/system-calls.js]: Catching up with the slack discussion. This should probably be refactored even before we start thinking about test, the error handling code is duplicated.

File: test/unit/adb-commands-specs.js (hunk 8 above)
Comment (sebv) [inline test/unit/adb-commands-specs.js]: @moizv you can set the mock return value, see above.

**Comment by moizjv** [comment]
@sebv refactored test to include mocks instead of stubs
